src/hitchhiker/tree/core.clj

```diff
@@ -301,174 +390,202 @@
-             (backtrack-up-path-until
-               path
-               (fn [parent index]
-                 (< (inc index) (count (:children parent)))))]
-    (let [next-index (-> common-parent-path peek inc)
-          parent (-> common-parent-path pop peek)
-          new-sibling (resolve (nth (:children parent) next-index))
-          ;; We must get back down to the data node
-          sibling-lineage (into []
-                                (take-while #(or (index-node? %)
-                                                 (data-node? %)))
-                                (iterate #(let [c (-> % :children first)]
-                                            (if (tree-node? c)
-                                              (resolve c)
-                                              c))
-                                         new-sibling))
-          path-suffix (-> (interleave sibling-lineage
-                                      (repeat 0))
-                          (butlast)) ; butlast ensures we end w/ node
-          ]
-      (-> (pop common-parent-path)
-          (conj next-index)
-          (into path-suffix)))))
-
-(defn forward-iterator
-  "Takes the result of a search and returns an iterator going
+  (go-try
+    (when-let [common-parent-path
+               (backtrack-up-path-until
+                path
+                (fn [parent index]
+                  (< (inc index) (count (:children parent)))))]
+      (let [next-index (-> common-parent-path peek inc)
+            parent (-> common-parent-path pop peek)
+            new-sibling (<? (resolve (nth (:children parent) next-index)))
+            ;; We must get back down to the data node
+            sibling-lineage (loop [res [new-sibling]
+                                    s new-sibling]
+                              (let [c (-> s :children first)
+                                    c (if (tree-node? c)
+                                        (<? (resolve c))
+                                        c)]
+                                (if (or (index-node? c)
+                                        (data-node? c))
+                                  (recur (conj res c) c)
+                                  res)))
+            ;; iterate cannot do blocking operations with core.async
+            #_(into []
+                  (take-while #(or (index-node? %)
+                                   (data-node? %)))
+                  (iterate #(let [c (-> % :children first)]
+                              (if (tree-node? c)
+                                (<?? S (resolve c S))
+                                c))
+                           new-sibling))
+            path-suffix (-> (interleave sibling-lineage
+                                        (repeat 0))
+                            (butlast)) ; butlast ensures we end w/ node
+            ]
+        (-> (pop common-parent-path)
+            (conj next-index)
+            (into path-suffix))))))
+
+
+
+#?(:clj
```

Comment:
Rather than locking this out of cljs, I'd love to have the forward-iterator take a channel to stream into, so that you can asynchronously iterate. By allowing the use to supply the channel, the user can control the amount of prefetching by specifying the size of the channel's buffer. That would be incredibly cool, and a powerful feature.